<launch>
  <arg name="standalone" default="false"/> <!-- debug nodes by setting then standalone -->
  <arg name="MANAGER" value="detic_detection_manager" unless="$(arg standalone)"/>
  <arg name="MANAGER" value="" if="$(arg standalone)"/>
  <arg name="LOAD_STATEMENT" value="load" unless="$(arg standalone)"/>
  <arg name="LOAD_STATEMENT" value="standalone" if="$(arg standalone)"/>
  <arg name="input_image" default="/kinect_head/rgb/image_rect_color"/>
  <arg name="input_depth" default="/kinect_head/depth_registered/image"/>
  <arg name="input_camera_info" default="/kinect_head/depth_registered/camera_info"/>
  <arg name="namespace" default="perception" />

  <arg name="_input_image" default="/$(arg namespace)/decompressed_image"/>
  <arg name="_input_depth" default="/$(arg namespace)/decompressed_depth"/>

  <group ns='$(arg namespace)'>

    <node name="$(arg MANAGER)" pkg="nodelet" type="nodelet" args="manager"/>

    <include file="$(find frmax_ros)/launch/decompress_depth.launch">
      <arg name="input_image" value="$(arg input_image)"/>
      <arg name="input_depth" value="$(arg input_depth)"/>
      <rosparam>
        queue_size: 100
      </rosparam>
    </include>

    <node pkg="nodelet" type="nodelet" name="decompress_points"
          args="$(arg LOAD_STATEMENT) depth_image_proc/point_cloud_xyzrgb $(arg MANAGER)">
      <remap from="rgb/camera_info" to="$(arg input_camera_info)"/>
      <remap from="rgb/image_rect_color" to="$(arg _input_image)"/>
      <remap from="depth_registered/image_rect" to="$(arg _input_depth)"/>
      <rosparam>
        queue_size: 100
      </rosparam>
    </node>

    <node name="tf_transform"
          pkg="nodelet" type="nodelet"
          args="$(arg LOAD_STATEMENT) jsk_pcl_utils/TfTransformCloud $(arg MANAGER)">
      <remap from="~input" to="depth_registered/points"/>
      <rosparam>
        target_frame_id: base_footprint
      </rosparam>
    </node>

    <node name="boxfiler" pkg="nodelet" type="nodelet"
      args="standalone jsk_pcl/AttentionClipper"
      output="screen">
      <remap from="~input/points" to="tf_transform/output"/>
      <rosparam>
          use_multiple_attention: false
          dimension_x: 1.0
          dimension_y: 0.8
          dimension_z: 0.4
          initial_pos: [0.5, 0.0, 0.95]
      </rosparam>
      <param name="frame_id" value="base_footprint" />
    </node>

    <node name="ExtractIndices" pkg="nodelet" type="nodelet"
      args="standalone jsk_pcl/ExtractIndices"
      output="screen">
      <remap from="~input" to="tf_transform/output"/>
      <remap from="~indices" to="boxfiler/output/point_indices"/>
      <rosparam>
        keep_organized: true
      </rosparam>
    </node>

    <node name="hsi_filter" pkg="nodelet" type="nodelet"
      args="standalone jsk_pcl/HSIColorFilter" 
      output="screen">
      <remap from="~input" to="/perception/ExtractIndices/output" />
      <rosparam>
        use_indices: false
        keep_organized: false
        h_limit_max: 127
        h_limit_min: 18
      </rosparam>
    </node>
    <node name="pose_detector" pkg="frmax_ros" type="cup_pose_publisher.py" output="screen"/>
  </group>
</launch>
